### PR TITLE
[ Rfc ]Coercing sliced array to Python  object

### PR DESCRIPTION
**Pull request by demiurg337**
fix for https://github.com/cython/cython/issues/2159
(I started it earlier and had to finish it)

Potentially I went to wrong way, ready to comments.

### CODE REVIEW

**Comment by demiurg337** [comment]
test

**Comment by demiurg337** [comment]
up

File: Cython/Compiler/PyrexTypes.py [hunk 1]
```diff
@@ -2592,6 +2592,46 @@ def find_cpp_operation_type(self, operator, operand_type=None):
...
+        }
+        env.use_utility_code(CythonUtilityCode.load(
+            "carray_slice.to_py", "CConvert.pyx",
+            outer_module_scope=env.global_scope(),  # need access to types declared in module
+            context=context, compiler_directives=dict(env.global_scope().directives)))
+        self.to_tuple_function = to_tuple_function
+        self.to_py_function = to_py_function
+        return True
+
+    def to_py_call_code(self, source_code, start_code, stop_code, result_code, result_type, to_py_function=None):
```
Comment (da-woods) [inline Cython/Compiler/PyrexTypes.py:2621]: This looks like it breaks the general interface for `to_py_call_code`.

File: Cython/Compiler/ExprNodes.py [hunk 2]
```diff
@@ -13192,6 +13200,16 @@ def analyse_types(self, env):
 
     def generate_result_code(self, code):
         code.putln('%s; %s' % (
+            self.arg.type.to_py_call_code(
```
Comment (da-woods) [inline Cython/Compiler/ExprNodes.py:13203]: Can you ever end up with `arg` as something that isn't a `SliceIndexNode`? For example:

    a = some_array[1:10]
    return a

Could `a` pick up the type from the indexing operation, but the coercion would happen to a `NameNode`?

File: Cython/Compiler/PyrexTypes.py [hunk 3]
```diff
@@ -2592,6 +2592,46 @@ def find_cpp_operation_type(self, operator, operand_type=None):
...
+
+        safe_typename = self.base_type.specialization_name()
+
+        to_py_function = "__Pyx_carray_slice_to_py_%s" % safe_typename
+        to_tuple_function = "__Pyx_carray_slice_to_tuple_%s" % safe_typename
+
+        from .UtilityCode import CythonUtilityCode
+        context = {
+            'cname': to_py_function,
+            'to_tuple_cname': to_tuple_function,
```
Comment (scoder) [inline Cython/Compiler/PyrexTypes.py:2610]: Both functions seem symmetric, so they should use similar names as well.

File: Cython/Compiler/PyrexTypes.py (hunk 1 above)
Comment (scoder) [reply Cython/Compiler/PyrexTypes.py:2621]: Yes. It seems to me that going through an implementation tree node would be better than going through the type system for this. A sliced array isn't really a type by itself.

File: Cython/Compiler/PyrexTypes.py [hunk 4]
```diff
@@ -2592,6 +2592,46 @@ def find_cpp_operation_type(self, operator, operand_type=None):
...
+            outer_module_scope=env.global_scope(),  # need access to types declared in module
+            context=context, compiler_directives=dict(env.global_scope().directives)))
+        self.to_tuple_function = to_tuple_function
+        self.to_py_function = to_py_function
+        return True
+
+    def to_py_call_code(self, source_code, start_code, stop_code, result_code, result_type, to_py_function=None):
+        func = self.to_py_function if to_py_function is None else to_py_function
+
+        target_is_tuple = result_type.is_builtin_type and result_type.name == 'tuple'
```
Comment (scoder) [inline Cython/Compiler/PyrexTypes.py:2624]: Makes total sense to have these two versions, but the tuple case doesn't seem to be exercised in the tests.

File: Cython/Compiler/PyrexTypes.py (hunk 1 above)
Comment (demiurg337) [reply Cython/Compiler/PyrexTypes.py:2621]: @scoder  ok

File: Cython/Compiler/ExprNodes.py (hunk 2 above)
Comment (demiurg337) [reply Cython/Compiler/ExprNodes.py:13203]: I little bit forgot what is there, I know that I did it by potentially wrong way, comments will help to understand where are errors

File: Cython/Compiler/PyrexTypes.py (hunk 1 above)
Comment (demiurg337) [reply Cython/Compiler/PyrexTypes.py:2621]: ok, I will move

**Comment by scoder** [comment]
Hmm, hang on, I'm confused now. The previous implementation used Cython implemented utility functions. Why did you switch over to manual code generation? In fact, why not just change the existing `carray.to_py` conversion function to accept a start and end index?

**Comment by scoder** [comment]
Also, there's already a `carray_slicing.pyx` test. You can just add the new test there, no need for a new test file.

**Comment by demiurg337** [comment]
ok, thanks for the comments

**Comment by demiurg337** [comment]
I didn't right understand previous comment about movement code into separated node, now I see that I can simply enable utility for CArray type and to use them (if I right understood).

Thanks for the comments.

File: Cython/Compiler/ExprNodes.py [hunk 5]
```diff
@@ -5034,13 +5037,17 @@ def coerce_to(self, dst_type, env):
...
 
     def generate_result_code(self, code):
+        base_result = self.base.result()
+        result = self.result()
+
         if not self.type.is_pyobject:
-            error(self.pos,
-                  "Slicing is not currently supported for '%s'." % self.type)
+            if self.base.type.is_array:
+                code.putln("%s = %s;" % (result, base_result))
```
[the remark this answers is not in the document]
Comment (demiurg337) [reply Cython/Compiler/ExprNodes.py:5045]: If do you mean 
1) first to convert whole c-array to python list (typle, frozenset) ...
2) ... and then to use the new, created python list for creation of slice
ok, I will change !

[the remark this answers is not in the document]
Comment (scoder) [reply Cython/Compiler/ExprNodes.py:5045]: Hmm, no, I meant that the `SliceIndexNode` can do the copying of the array slice part itself, rather than having a separate node do it instead. It's the purpose of the `SliceIndexNode` to do the slicing, whatever the result is. And if the result is calculated by a copy function, so be it. But it's still up to the `SliceIndexNode` to call that function, right here in this place.

[the remark this answers is not in the document]
Comment (demiurg337) [reply Cython/Compiler/ExprNodes.py:5045]: ok, then clarified )

File: tests/run/carray_slicing.pyx [hunk 6]
```diff
@@ -275,3 +277,214 @@ def struct_ptr_iter():
...
+#    (25, 36, 49, 64, 81)
+#    set([64, 25, 36, 81, 49])
+#    frozenset([64, 25, 36, 81, 49])
+#    """
+#    cdef int i
+#    cdef int a[10]
+#    b = 5555
+#    for i in range(10):
+#        a[i] = i**2
+#    b = a
```
Comment (demiurg337) [inline tests/run/carray_slicing.pyx:426]: seems like bug for coercion c-array to py-types (similar occurs for slice)
Do I have to do fixes in this pull request?

File: tests/run/carray_slicing.pyx [hunk 7]
```diff
@@ -275,3 +277,214 @@ def struct_ptr_iter():
...
+#    >>> test_coercing_c_slice_by_casting()
+#    [25, 36, 49, 64, 81]
+#    (25, 36, 49, 64, 81)
+#    set([64, 25, 36, 81, 49])
+#    frozenset([64, 25, 36, 81, 49])
+#    """
+#    cdef int i
+#    cdef int a[10]
+     # is not initializaed
+#    #b = 5555
```
Comment (demiurg337) [inline tests/run/carray_slicing.pyx:478]: If "b" isn't initialized ...

File: tests/run/carray_slicing.pyx [hunk 8]
```diff
@@ -275,3 +277,214 @@ def struct_ptr_iter():
...
+#    for i in range(10):
+#        a[i] = i**2
+#    # error when initialization of new variable b
+#    b = a[1:5]
+#    # and for this variant also error
+#    b = a
```
Comment (demiurg337) [inline tests/run/carray_slicing.pyx:484]: There occurs error (for c array slice  too)
Is it bug?

**Comment by demiurg337** [comment]
Seems I have found some additional bugs for converting c-array (and slice of c-array) to py  (comments have been added)

Commits I will fix later.

File: Cython/Compiler/ExprNodes.py [hunk 9]
```diff
@@ -5026,23 +5029,59 @@ def coerce_to(self, dst_type, env):
...
+                    to_py_obj_type = 'frozenset'
+                elif dst_type.is_pyobject:
+                    to_py_obj_type = 'list'
+                else:
+                    error(self.pos,
+                          "Cannot convert slice of C array to '%s'" % dst_type)
+
+                self.type = dst_type
+                to_py_func_cname = "__Pyx_slice_carray_" + self.base.type.specialization_name()\
+                                        + '_to_py_' + to_py_obj_type
```
Comment (da-woods) [inline Cython/Compiler/ExprNodes.py:5049]: I think we normally prefer splitting lines with brackets than `\`, and string formatting to concatenation:

```
to_py_func_cname = "__Pyx_slice_carray_%s_to_py_%s" % (
      self.base.type.specialization_name(), to_py_obj_type)
```

Although don't copy my indenting because I never get it right

File: Cython/Utility/CConvert.pyx [hunk 10]
```diff
@@ -130,3 +130,69 @@ cdef inline tuple {{to_tuple_cname}}({{base_type}} *v, Py_ssize_t length):
         Py_INCREF(value)
         PyTuple_SET_ITEM(t, i, value)
     return t
+
+#################### c_array_slice.to_py ####################
```
Comment (da-woods) [inline Cython/Utility/CConvert.pyx:134]: Could you largely re-use the `carray.to_py` section? Your functions would just have to work out a start and end index (taking into account wraparound) and then call the existing convension functions?

I'm also not convinced we really need a set/frozenset version - these are probably used rarely enough that they could just go through a tuple.

File: tests/run/carray_slicing.pyx [hunk 11]
```diff
@@ -275,3 +276,227 @@ def struct_ptr_iter():
     return ([ value.i for value in my_structs[:5] ],
             [ ptr.i for ptr in my_structs[:5] ],
             [ inferred.i for inferred in my_structs[:5] ])
+
```
Comment (da-woods) [inline tests/run/carray_slicing.pyx:279]: I can't see any tests for wraparound although your code claims to handle it.

File: tests/run/carray_slicing.pyx [hunk 12]
```diff
@@ -275,3 +276,227 @@ def struct_ptr_iter():
...
+    """
+    cdef int i
+    cdef int a[10]
+    b = 5555
+    for i in range(10):
+        a[i] = i**2
+    b = a[8:]
+    return b
+
+cpdef list test_c_to_py_coercing_array_in_var_2():
```
Comment (da-woods) [inline tests/run/carray_slicing.pyx:342]: I don't see why specifying the return type here makes any difference to what the test does?

File: tests/run/carray_slicing.pyx [hunk 13]
```diff
@@ -275,3 +276,227 @@ def struct_ptr_iter():
...
+#
+#    # return lists
+#    #[0, 1, 4, 9, 16, 25, 36, 49, 64, 81]
+#    #[0, 1, 4, 9, 16, 25, 36, 49, 64, 81]
+#    #[0, 1, 4, 9, 16, 25, 36, 49, 64, 81]
+#    #[0, 1, 4, 9, 16, 25, 36, 49, 64, 81]
+
+
+# Error !!!!!!!!
+#def test_coercing_c_slice_by_casting():
```
Comment (da-woods) [inline tests/run/carray_slicing.pyx:454]: It looks like you've fixed this (above)? Can this commented out bit now go? Or is this different in some way?

Comment (demiurg337) [reply tests/run/carray_slicing.pyx:454]: These tests (commented) still fails
description of bug

b = 5555
# a can be c-array or slice of c-array
# b = a[1:5]
b = a

then when I am trying to cast

print(<list> b)
print(<tuple> b)
print(<set> b)
print(<frozenset> b)

result will be always the list, since dst_type will be always python object, not tuple_type, ... etc.


Second bug:
when b wasn't initialized before (like in previous code example)
b = a[1:5]
or
b = a
will raise the error

File: tests/run/carray_slicing.pyx (hunk 12 above)
Comment (demiurg337) [reply tests/run/carray_slicing.pyx:342]: ok, will change

File: Cython/Utility/CConvert.pyx (hunk 10 above)
Comment (demiurg337) [reply Cython/Utility/CConvert.pyx:134]: In https://github.com/cython/cython/pull/3151#discussion_r425810643 to support list, tuple, set, frozenset

Comment (da-woods) [reply Cython/Utility/CConvert.pyx:134]: I missed that comment - sorry!

Note that you can add use [`PySet_Add`](https://docs.python.org/3/c-api/set.html#c.PySet_Add) on frozenset's too - you don't have to go through a tuple.

File: Cython/Utility/CConvert.pyx [hunk 14]
```diff
@@ -130,3 +130,69 @@ cdef inline tuple {{to_tuple_cname}}({{base_type}} *v, Py_ssize_t length):
...
+
+        value = arr[i]
+        Py_INCREF(value)
+
+        {{if to_py_obj_type == 'list'}}
+        PyList_SET_ITEM(l, i - from_idx, value)
+        {{elif to_py_obj_type == 'tuple' or to_py_obj_type == 'frozenset'}}
+        PyTuple_SET_ITEM(l, i - from_idx, value)
+        {{elif to_py_obj_type == 'set'}}
+        PySet_Add(l, value)
```
Comment (da-woods) [inline Cython/Utility/CConvert.pyx:191]: Unlike `PyList_SET_ITEM` and `PyTuple_SET_ITEM` this can raise an exception. It also doesn't steal a reference, so I think you have a memory leak.

File: Cython/Utility/CConvert.pyx [hunk 15]
```diff
@@ -130,3 +130,69 @@ cdef inline tuple {{to_tuple_cname}}({{base_type}} *v, Py_ssize_t length):
...
+    l = PyTuple_New(length)
+    {{elif to_py_obj_type == 'set'}}
+    l = PySet_New(NULL)
+    {{endif}}
+
+    cdef object value
+    for i in range(<size_t> from_idx, <size_t> to_idx):
+
+        value = arr[i]
+        Py_INCREF(value)
```
Comment (da-woods) [inline Cython/Utility/CConvert.pyx:184]: I'm a bit worried about these two lines. I think `value` will be inferred to be `base_type`, and a new Python object will be created specifically to pass to `Py_INCREF`. Instead I think you should do `cdef object value` at the start of the function so that the conversion happens on `value = arr[i]`.

File: tests/run/carray_slicing.pyx (hunk 13 above)
Comment (da-woods) [reply tests/run/carray_slicing.pyx:454]: The first bug looks like an existing problem that probably doesn't need fixing here.

I think the second bug is not really a bug. If `b` isn't initialized then Cython gets it's type from the assignment. You could also do `cdef object b` to set the type of `b`.

File: tests/run/carray_slicing.pyx (hunk 7 above)
Comment (da-woods) [reply tests/run/carray_slicing.pyx:492]: See comment above - I don't think this is a bug. It just doesn't know that `b` should be an object.

File: Cython/Utility/CConvert.pyx (hunk 10 above)
Comment (demiurg337) [reply Cython/Utility/CConvert.pyx:134]: I used it before, but it fails in PyPy for frozenset, so I have used
__Pyx_PyFrozenSet_New
by line
```
{{if to_py_obj_type == 'frozenset'}}
l = frozenset(l)
{{endif}}
```
by converting from tuple

File: Cython/Utility/CConvert.pyx (hunk 14 above)
Comment (demiurg337) [reply Cython/Utility/CConvert.pyx:191]: ok, will fix

File: Cython/Utility/CConvert.pyx (hunk 10 above)
Comment (da-woods) [reply Cython/Utility/CConvert.pyx:134]: I've opened a [PyPy bug for this](https://foss.heptapod.net/pypy/pypy/-/issues/3251). You should probably keep your workaround (with appropriate reference counting fixes) rather than wait for a fix from PyPy.

File: Cython/Utility/CConvert.pyx (hunk 15 above)
Comment (demiurg337) [reply Cython/Utility/CConvert.pyx:184]: ```cdef object value```  is added before ``` for i in range( ...```, or do you mean move it at the begin of function?

Comment (demiurg337) [reply Cython/Utility/CConvert.pyx:184]: ```cdef object value``` exist before ```for i in range(<size_t> ...``` or do you mean to move it at the begin of function?

Comment (da-woods) [reply Cython/Utility/CConvert.pyx:184]: Sorry. You're right - you have `cdef object value` and I didn't see it. This should be OK where it is.

File: Cython/Utility/CConvert.pyx [hunk 16]
```diff
@@ -130,3 +130,69 @@ cdef inline tuple {{to_tuple_cname}}({{base_type}} *v, Py_ssize_t length):
...
+    int __Pyx_is_valid_index(Py_ssize_t i, Py_ssize_t limit)
+
+    list PyList_New(Py_ssize_t size)
+    void PyList_SET_ITEM(object p, Py_ssize_t pos, object o)
+
+    tuple PyTuple_New(Py_ssize_t size)
+    void PyTuple_SET_ITEM(object p, Py_ssize_t pos, object o)
+
+    set PySet_New(void *)
+    void PySet_Add(object p, object o)
```
Comment (da-woods) [inline Cython/Utility/CConvert.pyx:149]: Change this to `int PySet_Add(object p, object o) except -1` and Cython should handle the exception case.

File: tests/run/carray_slicing.pyx (hunk 11 above)
Comment (da-woods) [reply tests/run/carray_slicing.pyx:278]: It looks like the tests you've added are for boundschecking (which is good - this should be tested). You should probably test `a[:-5]` or `a[-2:]` or `a[-5:-2]` or `a[-5:8]` etc.

File: tests/run/carray_slicing.pyx [hunk 17]
```diff
@@ -275,3 +275,194 @@ def struct_ptr_iter():
...
+    Traceback (most recent call last):
+        ...
+    IndexError: Some of slice indexes is out of range
+    """
+    cdef int i
+    cdef int a[10]
+    b = 0
+    for i in range(10):
+        a[i] = i**2
+    return a[:11]
```
Comment (da-woods) [inline tests/run/carray_slicing.pyx:453]: Not 100% sure about this. By analogy with Python types like `list` this should probably work and just return the full array. Although it does feel dodgy to index off the end of an array.

File: tests/run/carray_slicing.pyx [hunk 18]
```diff
@@ -275,3 +275,194 @@ def struct_ptr_iter():
...
+    Traceback (most recent call last):
+        ...
+    IndexError: Wrong range of slice
+    """
+    cdef int i
+    cdef int a[10]
+    b = 0
+    for i in range(10):
+        a[i] = i**2
+    return a[3:2]
```
Comment (da-woods) [inline tests/run/carray_slicing.pyx:468]: By analogy with Python types like list this should probably just return an empty list.

File: tests/run/carray_slicing.pyx (hunk 17 above)
Comment (demiurg337) [reply tests/run/carray_slicing.pyx:453]: yes, python return full array, will fix later

**Comment by demiurg337** [comment]
Changes by comments were done

File: Cython/Compiler/ExprNodes.py [hunk 19]
```diff
@@ -5100,6 +5141,21 @@ def generate_result_code(self, code):
                     start_code,
                     stop_code,
                     code.error_goto_if_null(result, self.pos)))
+        elif self.base.type.is_array:
+            if self.to_py_func:
+                code.putln('%s = %s(%s, %s, %s, %s, %d, %d); %s' % (
+                    result,
+                    self.to_py_func,
+                    base_result,
+                    self.base.type.size,
```
Comment (scoder) [inline Cython/Compiler/ExprNodes.py:5150]: Do we make sure anywhere that we know the size of the array?

File: Cython/Compiler/ExprNodes.py [hunk 20]
```diff
@@ -10446,13 +10502,14 @@ def analyse_types(self, env):
...
         from_py = self.operand.type.is_pyobject
+
         if from_py and not to_py and self.operand.is_ephemeral():
             if not self.type.is_numeric and not self.type.is_cpp_class:
                 error(self.pos, "Casting temporary Python object to non-numeric non-Python type")
         if to_py and not from_py:
             if self.type is bytes_type and self.operand.type.is_int:
                 return CoerceIntToBytesNode(self.operand, env)
-            elif self.operand.type.can_coerce_to_pyobject(env):
+            elif self.operand.type.can_coerce_to_pyobject(env) or isinstance(self.operand, SliceIndexNode):
```
Comment (scoder) [inline Cython/Compiler/ExprNodes.py:10512]: This condition looks fishy. Can we really always do the coercion whenever we find a `SliceIndexNode`? Does it no longer depend on the operand type in that case?

File: Cython/Utility/CConvert.pyx [hunk 21]
```diff
@@ -130,3 +130,74 @@ cdef inline tuple {{to_tuple_cname}}({{base_type}} *v, Py_ssize_t length):
...
+    l = PyTuple_New(length)
+    {{elif to_py_obj_type == 'set'}}
+    l = PySet_New(NULL)
+    {{endif}}
+
+    cdef object value
+    if length > 0:
+        for i in range(<size_t> from_idx, <size_t> to_idx):
+            value = arr[i]
+            Py_INCREF(value)
```
Comment (scoder) [inline Cython/Utility/CConvert.pyx:187]: I would move this line down into the list/tuple cases and only do it when needed, i.e. not for the `set` case.

File: Cython/Utility/CConvert.pyx [hunk 22]
```diff
@@ -130,3 +130,74 @@ cdef inline tuple {{to_tuple_cname}}({{base_type}} *v, Py_ssize_t length):
...
+    l = PyList_New(length)
+    {{elif to_py_obj_type == 'tuple' or to_py_obj_type == 'frozenset'}}
+    l = PyTuple_New(length)
+    {{elif to_py_obj_type == 'set'}}
+    l = PySet_New(NULL)
+    {{endif}}
+
+    cdef object value
+    if length > 0:
+        for i in range(<size_t> from_idx, <size_t> to_idx):
```
Comment (scoder) [inline Cython/Utility/CConvert.pyx:185]: You are mixing signed (`i`) and unsigned types here. That's rarely a good idea in C. Just use `Py_ssize_t` for everything.

File: tests/run/carray_slicing.pyx (hunk 17 above)
Comment (scoder) [reply tests/run/carray_slicing.pyx]: > Although it does feel dodgy to index off the end of an array.

It's ok for the non-constant case.

File: Cython/Utility/CConvert.pyx [hunk 23]
```diff
@@ -130,3 +130,74 @@ cdef inline tuple {{to_tuple_cname}}({{base_type}} *v, Py_ssize_t length):
...
+    if wraparound:
+        if from_idx < 0: 
+            from_idx += arr_size
+
+        if to_idx < 0: 
+            to_idx += arr_size
+
+    if boundscheck:
+        if from_idx > to_idx:
+             from_idx = to_idx
```
Comment (scoder) [inline Cython/Utility/CConvert.pyx:165]: There is no guarantee that this makes `from_idx` and `to_idx` >= 0. What if `to_idx` is -2000 for an array of size 5? Then it's still negative after the wrap-around above, and so will be `from_idx`.

File: Cython/Utility/CConvert.pyx [hunk 24]
```diff
@@ -130,3 +130,74 @@ cdef inline tuple {{to_tuple_cname}}({{base_type}} *v, Py_ssize_t length):
...
         PyTuple_SET_ITEM(t, i, value)
     return t
+
+#################### c_array_slice.to_py ####################
+
+cdef extern from *:
+    void Py_INCREF(object o)
+    void Py_DECREF(object o)
+    object PyErr_Format(exc, const char *format, ...)
+    int __Pyx_is_valid_index(Py_ssize_t i, Py_ssize_t limit)
```
Comment (scoder) [inline Cython/Utility/CConvert.pyx:140]: Unused?

File: tests/run/carray_slicing.pyx [hunk 25]
```diff
@@ -275,3 +275,186 @@ def struct_ptr_iter():
...
+    print(<tuple> a[5:])
+    b = <set> a[5:]
+    assert isinstance(b, set)
+    print(sorted(b))
+    b = <frozenset> a[5:]
+    assert isinstance(b, frozenset)
+    print(sorted(b))
+
+
+@cython.wraparound(True)
```
Comment (scoder) [inline tests/run/carray_slicing.pyx:425]: Since this setting is the default, it doesn't add anything to the test. However, neither `@wraparound(False)` nor `@boundscheck(False)` nor both together are tested.

File: Cython/Utility/CConvert.pyx (hunk 10 above)
Comment (scoder) [reply Cython/Utility/CConvert.pyx:134]: You could also just declare `__Pyx_PyFrozenSet_New(PyObject* it)` and call it as `__Pyx_PyFrozenSet_New(NULL)`. It seems to support that. (You would then have to call `use_utility_code()` for the `pyfrozenset_new` utility in the `frozenset` case.)

File: Cython/Compiler/ExprNodes.py (hunk 19 above)
Comment (demiurg337) [reply Cython/Compiler/ExprNodes.py:5151]:
```
elif self.base.type.is_array:
```
- means that it is carray (CArrayType) which has required parameter size. 
(init function of CArrayType)
```
 def __init__(self, base_type, size): 
 super(CArrayType, self).__init__(base_type) 
 self.size = size 
```
and value of this parameter became value of attribute self.base.type.size. So self.base.type.size must be valid.

**Comment by demiurg337** [comment]
I tryed __Pyx_PyFrozenSet_New(NULL) with PySet_Add and I had error, so I simple used __Pyx_PyFrozenSet_New.
Changes by comments were done.

File: Cython/Compiler/ExprNodes.py (hunk 20 above)
Comment (demiurg337) [reply Cython/Compiler/ExprNodes.py]: Additional condition has been added

File: Cython/Utility/CConvert.pyx (hunk 23 above)
Comment (demiurg337) [reply Cython/Utility/CConvert.pyx]: I did 
```
from_idx = to_idx = 0
```

File: Cython/Utility/CConvert.pyx (hunk 24 above)
Comment (demiurg337) [reply Cython/Utility/CConvert.pyx]: removed

**Comment by demiurg337** [comment]
up

**Comment by demiurg337** [comment]
When __Pyx_PyFrozenSet_New(NULL) is used, occurs error
...
SystemError: ../Objects/setobject.c:2334: bad argument to internal function
when I have checked source of Python, error is in check
```
if (!PySet_Check(a) &&
        (!PyFrozenSet_Check(a) || Py_REFCNT(a) != 1)) {
        PyErr_BadInternalCall();
}
```
more precisely in
 Py_REFCNT(a) != 1.

Count of refferences of new frozenset object which is returned by __Pyx_PyFrozenSet_New(NULL) is more than one (in my case 22). It occurs since frozenset.tp_new for empty set returns empty frozenset-singleton.

So seems like  __Pyx_PyFrozenSet_New(NULL)  can't be used.

**Comment by scoder** [comment]
Note that `__Pyx_PyFrozenSet_New(NULL)` is not from CPython but from Cython, where it explicitly supports `NULL`:
https://github.com/cython/cython/blob/aa8b0d9b89030c0b9ace5698ae5131c623acefa4/Cython/Utility/Builtins.c#L444-L445

**Comment by demiurg337** [comment]
I saw it, and I meant the last part of function
```c
#if CYTHON_USE_TYPE_SLOTS
    return PyFrozenSet_Type.tp_new(&PyFrozenSet_Type, $empty_tuple, NULL);
#else
    return PyObject_Call((PyObject*)&PyFrozenSet_Type, $empty_tuple, NULL);
#endif
}
```
more precisely
PyFrozenSet_Type.tp_new(&PyFrozenSet_Type, $empty_tuple, NULL);
-- this function returns frozenset-singleton (with ref-count 22), and if this singleton is used by PySet_Add, it creates error which is mentioned in my previous comment.

**Comment by scoder** [comment]
Ah, ok, understood. Then you were right before and we're back to first creating a (non-frozen) set or a tuple initially and then converting it to a `frozenset` afterwards. Creating a `set` instead of a `tuple` should be more (memory) efficient in the face of duplicate values. Converting a `set` into a `frozenset` is just a flat copy of the internal data structure, with all the hashing and deduplication already done.

**Comment by scoder** [comment]
BTW, @demiurg337, everything that you tried and that failed along the way and required you to take a different route with a special case, would best be remembered as a code comment in a suitable place. Otherwise, in a few months or years, someone will look at this code and say "there's room for optimisation", where we already knew at some point that there is none.

**Comment by demiurg337** [comment]
Comments were added, set is used for creation frozenset now.

Have I to merge the commits?
(I see possible structure like:
-- tests for carray slicing 
-- test for c-array slice auto encoding 
-- code for coercing slice of c-arrays to python object)

About bug with frozenset, issue has been openned
https://foss.heptapod.net/pypy/pypy/-/issues/3251

File: Cython/Utility/CConvert.pyx [hunk 26]
```diff
@@ -132,3 +132,79 @@ cdef inline tuple {{to_tuple_cname}}({{base_type}} *v, Py_ssize_t length):
...
+    l = PyTuple_New(length)
+
+    # The PySet_New is used for creating frozenset,
+    # since PyFrozenSet_New has a bug with PySet_Add in PyPy.
+    #
+    # __Pyx_PyFrozenSet_New(NULL) doesn't work for this goal,
+    # since it returns immutable frozenset-singleton.
+    {{elif to_py_obj_type == 'set' or to_py_obj_type == 'frozenset'}}
+    l = PySet_New(NULL)
+    {{endif}}
```
Comment (scoder) [inline Cython/Utility/CConvert.pyx:188]: Just to be safe:
```suggestion
    {{else}}
        {{py: assert False, "Invalid target type for C array conversion: %r" % to_py_obj_type"}}
    {{endif}}
```

File: Cython/Utility/CConvert.pyx [hunk 27]
```diff
@@ -132,3 +132,79 @@ cdef inline tuple {{to_tuple_cname}}({{base_type}} *v, Py_ssize_t length):
...
+
+    cdef object value
+    if length > 0:
+        for i in range(<Py_ssize_t> from_idx, <Py_ssize_t> to_idx):
```
Comment (scoder) [inline Cython/Utility/CConvert.pyx:192]: These are typed already.
```suggestion
        for i in range(from_idx, to_idx):
```

File: tests/run/carray_slicing.pyx [hunk 28]
```diff
@@ -275,3 +275,185 @@ def struct_ptr_iter():
...
+    print(<tuple> a[5:])
+    b = <set> a[5:]
+    assert isinstance(b, set)
+    print(sorted(b))
+    b = <frozenset> a[5:]
+    assert isinstance(b, frozenset)
+    print(sorted(b))
+
+
+def test_coercing_c_slice_out_of_range():
```
Comment (scoder) [inline tests/run/carray_slicing.pyx:425]: Please also test with `PY_SSIZE_T_MIN` and `PY_SSIZE_T_MAX` as boundary indices (`cdef extern from *`). See, for example, the [unicode_slicing](https://github.com/cython/cython/blob/master/tests/run/unicode_slicing.pyx) tests. Code like `length = to_idx - from_idx` may overflow in C for sufficiently large negative values.

File: Cython/Utility/CConvert.pyx [hunk 29]
```diff
@@ -132,3 +132,79 @@ cdef inline tuple {{to_tuple_cname}}({{base_type}} *v, Py_ssize_t length):
...
+            if from_idx < 0:
+                from_idx = 0
+
+            if to_idx > arr_size:
+                to_idx = arr_size
+
+    length = to_idx - from_idx
+
+    {{if to_py_obj_type == 'list'}}
+    l = PyList_New(length)
```
Comment (scoder) [inline Cython/Utility/CConvert.pyx:177]: Could you rename this to `container`? `l` tends to look a lot like `1` or `I` in some fonts and is generally difficult to spot as a single character variable name, which makes the code less easy to read.
( `i` as an integer loop index variable is good since it plays in a different league.)

**Review by scoder — COMMENTED** [review]
Nice, I think we're almost there now.

File: Cython/Utility/CConvert.pyx (hunk 29 above)
Comment (demiurg337) [reply Cython/Utility/CConvert.pyx:177]: ok

File: Cython/Utility/CConvert.pyx [hunk 30]
```diff
@@ -162,7 +162,7 @@ cdef inline {{to_py_obj_type}} {{cname}}({{base_type}} *arr, Py_ssize_t arr_size
             to_idx += arr_size
 
     if boundscheck:
-        if from_idx > to_idx:
+        if from_idx > to_idx or from_idx == to_idx:
```
Comment (scoder) [inline Cython/Utility/CConvert.pyx:165]:
```suggestion
        if from_idx >= to_idx:
```

File: tests/run/carray_slicing.pyx [hunk 31]
```diff
@@ -452,6 +462,15 @@ def test_coercing_c_slice_out_of_range():
...
 
+    b = a[:SSIZE_T_MAX]
+    print(b)
+
+    b = a[SSIZE_T_MIN:]
+    print(b)
+
+    b = a[SSIZE_T_MIN:SSIZE_T_MAX]
+    print(b)
+
```
Comment (scoder) [inline tests/run/carray_slicing.pyx:478]: Some more tests:
```suggestion
    b = a[SSIZE_T_MIN:SSIZE_T_MAX]
    print(b)

    b = a[SSIZE_T_MAX:SSIZE_T_MIN]
    assert len(b) == 0

    b = a[SSIZE_T_MAX:]
    assert len(b) == 0

    b = a[SSIZE_T_MAX:None]
    assert len(b) == 0

    b = a[:SSIZE_T_MIN]
    assert len(b) == 0

    b = a[None:SSIZE_T_MIN]
    assert len(b) == 0

    b = a[:]
    print(b)

```

Comment (demiurg337) [reply tests/run/carray_slicing.pyx:478]: ok

**Comment by demiurg337** [comment]
Changes by comments were done

**Comment by scoder** [comment]
Looks like travis didn't pick up the latest change. I'll close and reopen to trigger a build.

**Comment by demiurg337** [comment]
I have merged the commits (maybe too early)
